enterprise/cmd/frontend/internal/auth/openidconnect/config.go

```
@@ -6,49 +6,54 @@ import (
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
-	"net/http"
 
-	"github.com/inconshreveable/log15"
 	"github.com/sourcegraph/log"
 
 	"github.com/sourcegraph/sourcegraph/cmd/frontend/auth/providers"
 	"github.com/sourcegraph/sourcegraph/enterprise/internal/licensing"
 	"github.com/sourcegraph/sourcegraph/internal/conf"
 	"github.com/sourcegraph/sourcegraph/internal/conf/conftypes"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
 	"github.com/sourcegraph/sourcegraph/schema"
 )
 
-var mockGetProviderValue *provider
+var mockGetProviderValue *Provider
 
-// getProvider looks up the registered openidconnect auth provider with the given ID.
-func getProvider(id string) *provider {
+// getProvider looks up the registered OpenID Connect authentication provider
+// with the given ID.
+func getProvider(id string) *Provider {
 	if mockGetProviderValue != nil {
 		return mockGetProviderValue
 	}
-	p, _ := providers.GetProviderByConfigID(providers.ConfigID{Type: providerType, ID: id}).(*provider)
+	p, _ := providers.GetProviderByConfigID(
+		providers.ConfigID{
+			Type: providerType,
+			ID:   id,
+		},
+	).(*Provider)
 	return p
 }
 
-func handleGetProvider(ctx context.Context, w http.ResponseWriter, id string) (p *provider, handled bool) {
-	handled = true // safer default
-
+// GetProviderAndRefresh retrieves the authentication provider with the given
+// type and ID, and refreshes the provider.
+func GetProviderAndRefresh(ctx context.Context, id string, getProvider func(id string) *Provider) (p *Provider, safeErrMsg string, err error) {
```

Instead of directly writing to the response stream, which hijacks the control from the call site, it should just return information that the call site needs.

---

Why not return a specific error type that has a `UserReadableMessage() (string, bool)` method vs. returning 3 values?

---

> Why not return a specific error type that has a `UserReadableMessage() (string, bool)` method vs. returning 3 values?

This is a good idea :) Worth as a followup I think, but here I'm simply [following existing patterns](https://sourcegraph.sourcegraph.com/search?q=context%3Aglobal+repo%3A%5Egithub%5C.com%2Fsourcegraph%2Fsourcegraph%24+safeErrMsg+string%2C+err+error&patternType=literal&groupBy=path) since it is not directly related to the intended changes.

---

Alright, fair enough!

---

What do we mean by "refresh the provider"?

---

Admittedly I was lazy to look up what "refresh" actually does in OIDC, will rephrase with an update :)